### ISSUE TITLE
Giving prior='log-uniform' to Integer dimension

### ISSUE THREAD

**Issue by frederikfaye** [scikit-optimize/scikit-optimize#735]
**TL;DR:** I would like to sample logarithmically over an Integer Dimension - is this possible? If not, is there an easy hack?

I would like to logarithmically uniformly sample my variable `batch_size`, which is an integer,  from a distribution over, say, (128,1024). A hack would be to sample another integer 'log2_batch_size' over (7,10), but what if I wanted sampling from in between the values of [128,512,256,1024] as well?

**Comment by homarb** [comment]
Would this work?
```python
Categorical(np.logspace(1,10, num=10, base=2, dtype=np.int)
```

**Comment by mfinno** [comment]
I think this (hack) is a bit better:

`Real(128, 1024, name='batch_size', prior='log-uniform')`

You would need to cast the output as int, but this method will cover the entire range in a log-uniform way. Also `Categorical` does not work with sckit-opt plots.

**Comment by frederikfaye** [comment]
@homarb I don't think that is desirable; beside the reasons @mfinno gives, I don't think gp_minimize works as well with categoricals, as the distance function is far less informative (if someone could confirm this, I would appreciate it). 

@mfinno, I like your idea, but I'm leaving the issue open to see if anyone want to implement this functionality. Btw, https://github.com/scikit-optimize/scikit-optimize/pull/675#issue-187282057 should solve the plotting of categorical dimensions.

**Comment by liebscher** [comment]
I just ran into this issue as well. For example, should we want to search over n_estimators for a random forest, the parameter only takes integer values (thus excluding a Real space). However, it seems overly hack-y to code this as Categoricals. Additionally, using `np.logspace` would violate the notion that this function is continuous. 

It might seem more roundabout and perhaps pedantic, but I suggest that a decent solution to this would be: given base 2 high and low values, log2 transform them, then choose random integers in [log2(min), log2(max)], then transform them back to linear space. I believe this would keep everything as integers and eliminate the need to do any casting.

**Comment by frederikfaye** [comment]
@liebscher Yep, that is also what I ended up doing, but that can only yield sampled values that are powers of 2, even though there are integers in between.

**Comment by liebscher** [comment]
@frederikfaye I'm no expert, so feel free to counter. However, I don't think there are any integers in between values, as you put it. Take for example the series [1, 2, 4, 8, 16, 32]. This is the logspace we ideally are concerned with, where each value is given equal likelihood in the distribution. I suppose by "in between" integers, you're thinking of say 11 or 12, which would seem to fall logarithmically as integers in that series. However, these two values don't fall in this log series, for exactly the fact that they aren't powers of 2. Should we ask for values that aren't base 2, then we would be changing the distribution we're sampling from. Additionally, suppose we were to get the integer that falls nearest the logarithmic scale trend (so 11 or 12), how do we treat the value that should be found between 1 and 2? Suppose we were to ignore that flaw and just grab 3, 6, 11, and so on. Then this is no longer a log scale (I don't have a proof of this but I don't believe there's a base that results in integers 2, 3, 4, 6, 8...). If we were to grab as many integers (and preserve equal likelihood as in the original distribution) as possible between numbers then we'd just have a uniform distribution!

Knowing this, I think it's perfectly reasonable to implement a prior on the Integer dimension that is of base 2 (or any other integer base the user may want to specify).

**Comment by frederikfaye** [comment]
@liebscher 
> I don't think there are any integers in between values

The idea is to cast the result of the sampling in log-space as an integer afterwards. In the case of log2:

Given some range `(13,1042)`, draw `r = numpy.random.uniform(low=np.log2(13), high=np.log2(1042))`, and then give back `sampled = 2 ** np.round(r).astype(np.integer)`.

`sampled` can be an integer that isn't in (2,4,8,16,...).

**Comment by liebscher** [comment]
@frederikfaye When I run:
```
r = np.random.uniform(low=np.log2(13), high=np.log2(1042), size=100000)
set(2 ** np.round(r).astype(np.integer))
```
the output is `{16, 32, 64, 128, 256, 512, 1024}`. Did you mean to place the rounding after the exponentiation?

In that case, I see what you're putting across. The output looks very similar to
```
cat = skopt.space.Real(1, 64, prior='log-uniform')
plt.hist(cat.rvs(n_samples=100000), bins=64);
```
I think we should consider implementing this prior on the Integer space. However, after thinking about this, in my own case and I think in others, I can't understand why we'd want to do this. As far as I know, there's no true evidence that using base 2 integers for certain parameters actually improves the model in any way except marginal performance improvements. What's the point in using a logspace if the results aren't base 2? It's certainly not to reduce the search space (since we set that explicitly with Bayesian search), as is the case normally with base 2 parameters.

**Comment by frederikfaye** [comment]
D'oh, sorry, I of course meant rounding after exponentiation!

> What's the point in using a logspace if the results aren't base 2?

I'm not quite sure what you mean. The point is the same when using `prior='log-uniform'` for `Real` dimensions; to sample over much larger intervals without having one end of the interval dominate the results. As to the "if the results aren't base 2"-part, why should we limit the parameter space in this way? It is quite unlikely that the best parameter is base 2, if all other integers in between are attainable.

**Comment by liebscher** [comment]
Of course, you're right! Well then, I say we implement this. I think it'd be valuable. @frederikfaye do you want to do this PR?

**Comment by frederikfaye** [comment]
@liebscher I'd love to at some point, but I don't really have time before New Year's. If you do, you're very welcome to try to implement it yourself. 

Thanks for all your constructive comments! :)

**Comment by janosh** [comment]
I'd also welcome this feature!

**Comment by holgern** [comment]
Should be fixed in 0.7.1
